channels_redis/serializers.py

```diff
@@ -0,0 +1,140 @@
+                raise ValueError(
+                    "symmetric_encryption_keys must be a list of possible keys"
+                )
+            try:
+                from cryptography.fernet import MultiFernet
+            except ImportError:
+                raise ValueError(
+                    "Cannot run with encryption without 'cryptography' installed."
+                )
+            sub_fernets = [self.make_fernet(key) for key in symmetric_encryption_keys]
+            self.crypter = MultiFernet(sub_fernets)
+        else:
+            self.crypter = None
+
+    @abc.abstractmethod
+    def dumps(self, message):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def loads(self, message):
+        raise NotImplementedError
+
+    def serialize(self, message):
+        """
+        Serializes message to a byte string.
+        """
+        message = self.dumps(message)
+        # ensure message is bytes
+        if isinstance(message, str):
+            message = message.encode("utf-8")
+        if self.crypter:
+            message = self.crypter.encrypt(message)
+
+        if self.random_prefix_length > 0:
+            # provide random prefix
+            message = (
+                random.getrandbits(8 * self.random_prefix_length).to_bytes(
+                    self.random_prefix_length, "big"
+                )
+                + message
+            )
+        return message
+
+    def deserialize(self, message):
+        """
+        Deserializes from a byte string.
+        """
+        if self.random_prefix_length > 0:
+            # Removes the random prefix
+            message = message[self.random_prefix_length :]  # noqa: E203
+
+        if self.crypter:
+            ttl = self.expiry if self.expiry is None else self.expiry + 10
+            message = self.crypter.decrypt(message, ttl)
+        return self.loads(message)
+
+
+class MissingSerializer(BaseMessageSerializer):
+    exception = None
+
+    def __init__(self, *args, **kwargs):
+        raise self.exception
+
+
+class JSONSerializer(BaseMessageSerializer):
+    dumps = staticmethod(json.dumps)
+    loads = staticmethod(json.loads)
+
+
+# code ready for a future in which msgpack may become an optional dependency
+try:
+    import msgpack
+except ImportError as exc:
+
+    class MsgPackSerializer(MissingSerializer):
+        exception = exc
+
+else:
+
+    class MsgPackSerializer(BaseMessageSerializer):
+        dumps = staticmethod(msgpack.packb)
+        loads = staticmethod(msgpack.unpackb)
+
+
+class SerializersRegistry:
+    def __init__(self):
+        self._registry = {}
+
+    def register_serializer(self, format, serializer_class):
+        """
+        Register a new serializer for given format
+        """
+        assert isinstance(serializer_class, type) and (
+            issubclass(serializer_class, BaseMessageSerializer)
+            or hasattr(serializer_class, "serialize")
+            and hasattr(serializer_class, "deserialize")
```

Comment:
Ok, I checked that black does not complain for this thus I can add it.

In my mind they were "useless" since I am very used to read boolean expressions and to know the precedence of operators. :smile:

Reply:
> ... know the precedence of operators...

Shouldn't have people need to stop and think here. +1 to the extra parentheses. 


Although... 🤔 Could we check `isinstance` on a `Protocol` here? (Perhaps not but...)